### PR TITLE
JAMES-2340 Alternative to additional matcher/mailet

### PR DESCRIPTION
**Pull request by chibenwa**
Alternative to https://github.com/linagora/james-project/pull/1336

Please review only https://github.com/linagora/james-project/commit/322723d7c5fc2feb978cb362d7a1d748ff0c7a6e

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java [hunk 1]
```diff
@@ -80,20 +80,26 @@ public SpamAssassinResult build() {
...
         }
     }
 
+    private final boolean isSpam;
     private final String hits;
     private final String requiredHits;
     private final Map<String, String> headersAsAttribute;
 
-    private SpamAssassinResult(String hits, String requiredHits, Map<String, String> headersAsAttribute) {
+    private SpamAssassinResult(boolean isSpam, String hits, String requiredHits, Map<String, String> headersAsAttribute) {
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java]: we should consider boolean as a code smell in general and use a type every time

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java [hunk 2]
```diff
@@ -111,9 +113,17 @@ private void querySpamAssassin(Mail mail, MimeMessage message, SpamAssassinInvok
...
                     .build(), recipient);
+
+            if (result.isSpam()) {
+                planDeliveryAsSpam(mail, recipient);
+            }
         }
     }
 
+    private void planDeliveryAsSpam(Mail mail, MailAddress recipient) throws UsersRepositoryException {
+        mail.setAttribute(MailStore.DELIVERY_PATH_PREFIX + usersRepository.getUser(recipient), DefaultMailboxes.SPAM);
```
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/SpamAssassin.java]: Great, I didn't know that.

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java (hunk 1 above)
Comment (aduprat) [reply server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java]: Maybe it's time to enhance this class:
 - cast _hits_ & _requiredHits_ to double in **Builder**
 - _isSpam_ is _hits_ >= _requiredHits_

Comment (aduprat) [reply server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java]: This modification has been reverted.

**Comment by chibenwa** [comment]
True... The *complex* design not implemented here (and more expensive) would have had been to have **per-recipient** attributes.

**Comment by chibenwa** [comment]
```
[30d3f48f5277edef7e28effac25814ba7a4d325a] Tests in error: 
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
[30d3f48f5277edef7e28effac25814ba7a4d325a]   SpamAssassinContract.setup:79 » DomainList spamer.com already exists.
``` 

In Cassandra integration tests

**Comment by aduprat** [comment]
`[INFO] Apache James :: Server :: JMAP :: Cassandra Integration testing FAILURE [56:34 min]`

```
Tests in error: 
  SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:190 » ConditionTimeout
  SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:237 » ConditionTimeout
  SpamAssassinContract.spamAssassinShouldForgetMessagesMovedOutOfSpamFolderUsingIMAP:423 » ConditionTimeout
```

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/ImmutableMailboxMessage.java [hunk 3]
```diff
@@ -88,23 +88,19 @@ public ImmutableMailboxMessage from(MailboxId mailboxId, MailboxMessage message)
...
     private final String mediaType;
     private final String subType;
     private final long bodyOctets;
     private final long fullContentOctets;
     private final long headerOctets;
     private final Long textualLineCount;
-    private final InputStream headerContent;
-    private final InputStream fullContent;
+    private final byte[] headerContent;
+    private final byte[] fullContent;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/ImmutableMailboxMessage.java]: That means we cary the double of what is needed, isn't it? (headerContent + bodyContent + fullContent)

Can't we manage to have something with an offset?

Comment (aduprat) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/ImmutableMailboxMessage.java]: Maybe something like _Bytes.concat_
but I would like to have Jenkins feedback on this pull request before optimizing things.

**Comment by aduprat** [comment]
:+1:

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**